### PR TITLE
Fix AttrProxy module reuse in Dynamo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185300

Dynamo checks side-effect tracking before VariableBuilder wraps a value, but proxy_tensor _AttrProxy modules are unwrapped inside wrap_module before they are tracked. When the same proxy-backed module is first seen as a local key and later through WeakKeyDictionary internals, the early lookup checks the proxy id, misses the real module that was already tracked, unwraps to the same base module, and then trips the duplicate mutation-tracking assertion.

Fix this by re-checking side-effect tracking immediately after _AttrProxy is unwrapped to its base module. The reuse path mirrors the existing duplicate-object handling: install a duplicate guard when possible and refresh the UnspecializedNNModuleVariable stack source before returning the existing tracker. This keeps the fix scoped to the proxy-module path instead of relaxing the global mutation-tracking assertion.

Test Plan:

python test/dynamo/test_dicts.py -k test_weakkeydict_attr_proxy_key

python test/dynamo/test_dicts.py -k weakref_dict

lintrunner -a

Benchmark Results:

Duplicate AttrProxy WeakKeyDictionary compile path: main failed with AssertionError after 3545.040 ms; fix-161053 passed in 1982.965 ms. Non-duplicate AttrProxy compile microbenchmark over 50 iterations: main median 15.008 ms, mean 15.084 ms; fix median 15.467 ms, mean 15.574 ms.

Fixes #161053

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I marked this PR ready for review, added the topic: not user facing label, and requested review from frgossen plus ydwu4 for the Dynamo AttrProxy code path.

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -2402,6 +2402,15 @@ def wrap_module(self, value: torch.nn.Module) -> VariableTracker:
                 # type: ignore[attr-defined]
                 value = value.get_base()
                 self.source = AttrProxySource(self.source)
+                if value in self.tx.output.side_effects:
```
Comment (bobrenjc93) [inline torch/_dynamo/variables/builder.py]: This block is a near-verbatim copy of the dedup logic in `_call_impl` (lines ~700-717): same `side_effects` lookup, same `make_dupe_guard`, same `set_nn_module_stack_source` refresh. Please factor it into a small shared helper (e.g. `_reuse_tracked_variable(value)`) so the two copies can't drift — in particular the `_call_impl` version guards the stack-source refresh with `isinstance(value, torch.nn.Module)` and carries a comment explaining why the nn_module_stack is updated; both are lost here. Also worth a one-line comment stating why the re-check is needed at all (the early check in `_call_impl` sees the proxy's id, not the base module's).

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -2402,6 +2402,15 @@ def wrap_module(self, value: torch.nn.Module) -> VariableTracker:
...
+                    side_effect_result = self.tx.output.side_effects[value]
+                    dup_guard = make_dupe_guard(self.source, side_effect_result.source)
+                    if dup_guard:
+                        self.install_guards(dup_guard)
+
+                    if isinstance(side_effect_result, UnspecializedNNModuleVariable):
+                        side_effect_result.set_nn_module_stack_source(self.source)
```
Comment (bobrenjc93) [inline torch/_dynamo/variables/builder.py]: The stack-source refresh is unconditional, so whichever source Dynamo happens to realize *second* wins. In the reported repro the second source is `AttrProxySource(WeakRefCallSource(ConstDictKeySource(...)))` (WeakKeyDictionary internals) while the first is the user-facing `AttrProxySource(LocalSource('key'))` — overwriting means `nn_module_stack` metadata for the base module becomes derived from dict internals and depends on wrapping order. Consider keeping the existing stack source when the new one is not user-facing, or at least noting this tradeoff in a comment.

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -2402,6 +2402,15 @@ def wrap_module(self, value: torch.nn.Module) -> VariableTracker:
                 # type: ignore[attr-defined]
                 value = value.get_base()
                 self.source = AttrProxySource(self.source)
+                if value in self.tx.output.side_effects:
+                    side_effect_result = self.tx.output.side_effects[value]
+                    dup_guard = make_dupe_guard(self.source, side_effect_result.source)
+                    if dup_guard:
```
Comment (bobrenjc93) [inline torch/_dynamo/variables/builder.py]: When `make_dupe_guard` returns None (e.g. one source is local and the other global) this returns the existing tracker with no guard installed at all on the new proxy source — not even the TYPE_MATCH that the normal path below would install. Unlike the `_call_impl` dedup, the two objects here are genuinely distinct Python objects (`AttrProxy._modules` mints a fresh proxy on every access), so nothing pins this source to the same base module on a later call. Consider falling through to the normal wrapping path (or installing TYPE_MATCH) when no dupe guard can be made.

File: test/dynamo/test_dicts.py [hunk 4]
```diff
@@ -908,6 +909,25 @@ def fn(x):
         x = torch.randn(4)
         self.assertEqual(fn(x), opt_fn(x))
 
+    def test_weakkeydict_attr_proxy_key(self):
```
Comment (bobrenjc93) [inline test/dynamo/test_dicts.py:991]: The regression test reaches into FX private internals (`_ModuleStackTracer`, `tracer.proxy_type`) and hand-builds an `_AttrProxy`, so it is fragile against refactors of proxy_tensor and doesn't cover the path that actually broke (non-strict `torch.export` of a module containing `torch.cond` with a shared submodule, per #161053). Please add an end-to-end test in the export/HOP suites that reproduces the reported failure, and reference the issue number in a comment here so the intent survives.

File: test/dynamo/test_dicts.py [hunk 5]
```diff
@@ -908,6 +909,25 @@ def fn(x):
...
+        class Root(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.child = torch.nn.Module()
+                self.alias = self.child
```
Comment (bobrenjc93) [inline test/dynamo/test_dicts.py]: `self.alias = self.child` has no effect on this test: `proxy_type` is set unconditionally in `_ModuleStackTracer.__init__`, and the shared-module detection only flips `enable_attr_proxy`, which isn't consulted because the proxy is constructed directly. Either drop it or add a comment saying it's there to mimic the real shared-submodule setup that makes AttrProxy engage.

**Review by bobrenjc93 — APPROVED** [review]

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py]: My agent says I factored the side-effect reuse logic into _reuse_tracked_variable; both the normal path and the post-AttrProxy-unwrapping path now share it, with a comment explaining why the second lookup is required.

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py]: My agent says dictionary-key-derived sources no longer replace nn_module_stack provenance. The focused test now forces the weak-dictionary lookup before a module call and asserts that the stack retains the user-facing key.get_base() path.

File: torch/_dynamo/variables/builder.py (hunk 3 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py]: My agent says the no-dupe-guard case now installs ID_MATCH guards on both sources, since a type-only guard would not preserve the alias relationship. A new local/global AttrProxy test changes the alias and verifies recompilation plus correct output.

File: test/dynamo/test_dicts.py (hunk 4 above)
Comment (jansel) [reply test/dynamo/test_dicts.py:991]: My agent says I added an end-to-end regression in test/export/test_export.py using non-strict torch.export, torch.cond, a shared module, and a WeakKeyDictionary closure. It reproduces the exact duplicate-tracking path on main without directly constructing FX tracer internals.

File: test/dynamo/test_dicts.py (hunk 5 above)
Comment (jansel) [reply test/dynamo/test_dicts.py]: My agent says I removed the ineffective alias from the direct unit test. The shared alias now appears only in the end-to-end export test, where it is required to enable AttrProxy behavior.

**Comment by jansel** [comment]
My agent says I rebased onto current main and updated the fix for the review feedback: reuse logic is centralized, local/global aliases receive sound identity guards, dictionary internals no longer replace user-facing module-stack provenance, and an end-to-end non-strict export/torch.cond regression now covers the issue. Local validation passed for all 325 dictionary tests, the related export/cond tests, and lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says 

> The six related CI failures all came from the nested-graph-break wrapper renaming the test's user-facing source from `L['key']` to `L['args'][1]`. The regression now expects the correct source in either mode. During review, a reverse-order alias case was also found and fixed: when an AttrProxy was tracked before its raw base module, stale state could be reused without recompiling. The full dictionary suite, both wrapped AttrProxy tests, related export/cond tests, ID_MATCH serialization tests, and `lintrunner -a` pass locally.
